### PR TITLE
Add support distinct() function of JSON Path

### PR DESCRIPTION
**Pull request by abudevich**

### CODE REVIEW

File: vividus-tests/src/main/resources/story/integration/JSON.story [hunk 1]
```diff
@@ -217,3 +217,24 @@ Then JSON element from `${json}` by JSON path `$.store` is equal to `
...
+  "name": [
+    "abc"
+  ],
+  "date": "2023-11-16"
+}
+` by JSON path `$.type.distinct()` is equal to `[
+                                   "a",
+                                   "b",
+                                   "c"
+                                ]` IGNORING_ARRAY_ORDER
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/JSON.story:240]:
```suggestion
` by JSON path `$.type.distinct()` is equal to `[
   "a",
   "b",
   "c"
]` IGNORING_ARRAY_ORDER
```

File: vividus/build.gradle [hunk 2]
```diff
@@ -37,7 +37,8 @@ tasks.register('systemTest', Test) {
 test {
     jvmArgs = [
             '--add-opens=java.base/java.util=ALL-UNNAMED',
-            '--add-opens=java.base/java.lang=ALL-UNNAMED'
+            '--add-opens=java.base/java.lang=ALL-UNNAMED',
```
Comment (valfirst) [inline vividus/build.gradle:40]: why is it needed?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus/build.gradle (hunk 2 above)
Comment (abudevich) [reply vividus/build.gradle:40]: Unit tests from `vividus:test` didn't pass without this

File: vividus-util/src/main/java/org/vividus/util/FieldUtils.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+
+/**
+ * Hack for updating final static field of external library
+ */
+public final class FieldUtils
```
[the remark this answers is not in the document]
Comment (abudevich) [reply vividus-util/src/main/java/org/vividus/util/FieldUtils.java:27]: do you mean to move this logic into `JsonPathUtils.class` or remove at all or another solution ?

File: vividus-tests/src/main/resources/story/integration/JSON.story [hunk 4]
```diff
@@ -217,3 +217,24 @@ Then JSON element from `${json}` by JSON path `$.store` is equal to `
   }
 }
 `
+
+Scenario: Validate using distinct() function in JsonPath
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/JSON.story:221]:
```suggestion
Scenario: Validate `distinct()` function of JsonPath
```

File: vividus-util/src/main/java/org/vividus/util/FieldUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (valfirst) [reply vividus-util/src/main/java/org/vividus/util/FieldUtils.java:27]: > move this logic into `JsonPathUtils`

File: vividus-util/src/main/java/org/vividus/util/json/JsonPathUtils.java [hunk 5]
```diff
@@ -74,6 +81,22 @@ public static void setJacksonConfiguration()
         Configuration.setDefaults(new JacksonConfiguration());
     }
 
+    public static void setDistinctPathFunction()
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/json/JsonPathUtils.java:84]:
```suggestion
    public static void addDistinctFunction()
```

File: vividus/build.gradle (hunk 2 above)
Comment (valfirst) [reply vividus/build.gradle:40]: what is the error? (I'm interested in this particular flag)

**Review by valfirst — COMMENTED** [review]
Also please make sure to add/update documentation

File: vividus/build.gradle (hunk 2 above)
Comment (abudevich) [reply vividus/build.gradle:40]:
```
testNoMatchedSteps(StdOut)
java.lang.NoClassDefFoundError: Could not initialize class org.vividus.util.json.JsonPathUtils
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at org.vividus.configuration.Vividus.init(Vividus.java:38)
	at org.vividus.runner.StepsCounter.main(StepsCounter.java:65)
	at org.vividus.runner.StepsCounterTests.testCounter(StepsCounterTests.java:194)
	at org.vividus.runner.StepsCounterTests.testNoMatchedSteps(StepsCounterTests.java:91)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.ExceptionInInitializerError [in thread "Test worker"]
	at org.vividus.util.json.JsonPathUtils.setDistinctPathFunction(JsonPathUtils.java:92)
	at org.vividus.util.json.JsonPathUtils.<clinit>(JsonPathUtils.java:47)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at org.vividus.configuration.Vividus.init(Vividus.java:38)
	at org.vividus.configuration.VividusTests.shouldDrawBannerAfterLoggerConfigurations(VividusTests.java:52)
	... 3 more
```

File: vividus-util/src/main/java/org/vividus/util/FieldUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (abudevich) [reply vividus-util/src/main/java/org/vividus/util/FieldUtils.java]: changed approach

File: vividus-plugin-json/build.gradle [hunk 6]
```diff
@@ -1,15 +1,13 @@
...
 dependencies {
     implementation project(':vividus-engine')
     implementation project(':vividus-reporter')
     implementation project(':vividus-soft-assert')
     implementation project(':vividus-util')
     api(group: 'net.javacrumbs.json-unit', name: 'json-unit', version: '3.2.2')
-    implementation(group: 'com.jayway.jsonpath', name: 'json-path', version: "${jsonPathVersion}")
-    implementation(group: 'com.jayway.jsonpath', name: 'json-path-assert', version: "${jsonPathVersion}")
+    implementation(group: 'com.github.vividus-framework.JsonPath', name: 'json-path', version: 'fd1bbfe6f9')
+    implementation(group: 'com.jayway.jsonpath', name: 'json-path-assert', version: '2.8.0')
```
Comment (valfirst) [inline vividus-plugin-json/build.gradle:10]: the versions should be aligned

Comment (valfirst) [reply vividus-plugin-json/build.gradle:10]: it seems the best way is to move the version to the root `build.gradle`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-json/build.gradle (hunk 6 above)
Comment (abudevich) [reply vividus-plugin-json/build.gradle]: updated, thanks )

File: build.gradle [hunk 7]
```diff
@@ -72,6 +72,8 @@ ext {
         powermock:              'a0be5c4bbe',
         // eyes and eyesUniversal versions should be updated and tested together
         eyes:                   '5.64.7',
-        eyesUniversal:          '5.63.13'
+        eyesUniversal:          '5.63.13',
+        jsonPath:               'fd1bbfe6f9',
+        jsonPathAssert:         '2.8.0'
```
Comment (valfirst) [inline build.gradle:77]: we should use a single version for all JsonPath artifacts

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 8]
```diff
@@ -5,6 +5,41 @@ The plugin provides a set of actions for transformation and validation of JSON d
...
+    "abc"
+  ],
+  "date": "2023-11-16"
+}
+` by JSON path `$.type.distinct()` is equal to `[
+   "a",
+   "b",
+   "c"
+]` IGNORING_ARRAY_ORDER
+----
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:41]: it should go in general section headed `Steps`

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 9]
```diff
@@ -5,6 +5,41 @@ The plugin provides a set of actions for transformation and validation of JSON d
 :plugin-short-name: json
 include::partial$plugin-installation.adoc[]
 
+== JSON Path
+
+Vividus uses inside https://github.com/json-path/JsonPath#readme[Jayway JsonPath] implementation.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:10]:
```suggestion
JSON path expressions always refer to a JSON structure in the same way as XPath expression are used in combination with an XML document. The "root member object" in JsonPath is always referred to as `$` regardless if it is an object or array.

JSON path expressions can use the dot–notation

[source]
----
$.store.book[0].title
----

or the bracket–notation

[source]
----
$['store']['book'][0]['title']
----

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 10]
```diff
@@ -5,6 +5,41 @@ The plugin provides a set of actions for transformation and validation of JSON d
...
+For building paths you can use standart https://github.com/json-path/JsonPath#operators[operators] and https://github.com/json-path/JsonPath#functions[functions].
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:11]:
```suggestion
Also the number of https://github.com/json-path/JsonPath#operators[operators] and https://github.com/json-path/JsonPath#functions[functions] are available for use in JSON paths out of the box. At the same time https://here_the_link_to_functions_section_should_go[additional functions] are available.
```

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 11]
```diff
@@ -5,6 +5,41 @@ The plugin provides a set of actions for transformation and validation of JSON d
...
+
+=== distinct()
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:13]:
```suggestion
=== JSON Path Functions
==== `distinct()`
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 12]
```diff
@@ -5,6 +5,56 @@ The plugin provides a set of actions for transformation and validation of JSON d
...
+    ]
+}
+----
+
+.distinct() function of JSON Path example
+|===
+|JSON Path |Result
+
+|`$.type.distinct()`
+|`["b","a","c"]`
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:55]: please provide more examples:
- array of objects
- what will happen if it's applied to object

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 13]
```diff
@@ -5,6 +5,56 @@ The plugin provides a set of actions for transformation and validation of JSON d
...
+
+=== JSON Path Functions
+
+==== `distinct()`
+
+Vividus provides ability to use additional function for filtering repeated values.
+
+* `distinct()` - Provides an array containing only unique items from the input array.
+
+[source,json]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:34]: provide description what it is

Comment (abudevich) [reply docs/modules/plugins/pages/plugin-json.adoc]: moved into table with examples

File: docs/modules/plugins/pages/plugin-json.adoc (hunk 12 above)
Comment (abudevich) [reply docs/modules/plugins/pages/plugin-json.adoc]: added example with array of objects
and for object - if I right understand, Jayway implementation by default remove repeating pairs name-value.
Example:
Screenshot 2023-12-05 at 16 36 03

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 14]
```diff
@@ -121,6 +121,104 @@ The steps syntax uses two internal (VIVIDUS-only) terms:
...
+$.store.book[0].title
+----
+
+or the bracket–notation
+
+[source]
+----
+$['store']['book'][0]['title']
+----
+
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:141]:
```suggestion
Also the number of https://github.com/json-path/JsonPath#operators[operators] and https://github.com/json-path/JsonPath#functions[functions] are available for use in JSON paths out of the box. At the same time there are additional functions listed below.


```

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 15]
```diff
@@ -121,6 +121,106 @@ The steps syntax uses two internal (VIVIDUS-only) terms:
...
+Also the number of https://github.com/json-path/JsonPath#operators[operators] and https://github.com/json-path/JsonPath#functions[functions] are available for use in JSON paths out of the box. At the same time there are additional functions listed below.
+
+==== JSON Path Functions
+
+===== `distinct()`
+
+Vividus provides ability to use additional function for filtering repeated values.
+
+* `distinct()` - Provides an array containing only unique items from the input array.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:150]:
```suggestion

* `distinct()` - removes duplicated items from the input array.
```

**Review by valfirst — APPROVED** [review]

**Pull request merged**